src/Aspire.Hosting/Extensions/ResourceBuilderExtensions.cs

```diff
@@ -228,6 +228,38 @@ public static IResourceBuilder<TDestination> WithReference<TDestination>(this IR
         });
     }
 
+    /// <summary>
+    /// Injects a connection string as an environment variable. The format of the environment variable will be "ConnectionStrings__{name}={value}." If the
+    /// connection string is not specified, the configuration system will be queried for a connection string using the connection string name.
+    /// </summary>
+    /// <typeparam name="TDestination"></typeparam>
+    /// <param name="builder">The resource where connection string will be injected.</param>
+    /// <param name="connectionString">A connection string</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{TDestination}"/>.</returns>
+    public static IResourceBuilder<TDestination> WithReference<TDestination>(this IResourceBuilder<TDestination> builder, ConnectionString connectionString)
+        where TDestination : IResourceWithEnvironment
+    {
+        var connectionStringName = $"{ConnectionStringEnvironmentName}{connectionString.Name}";
+
+        return builder.WithEnvironment(context =>
+        {
+            var connectionStringValue = connectionString.Value ??
+                builder.ApplicationBuilder.Configuration.GetConnectionString(connectionString.Name);
+
+            if (string.IsNullOrEmpty(connectionStringValue))
+            {
+                throw new DistributedApplicationException($"A connection string for '{connectionString.Name}' could not be retrieved.");
+            }
+
+            if (builder.Resource is ContainerResource)
+            {
+                connectionStringValue = HostNameResolver.ReplaceLocalhostWithContainerHost(connectionStringValue, builder.ApplicationBuilder.Configuration);
+            }
+
+            context.EnvironmentVariables[connectionStringName] = connectionStringValue;
```

Comment:
```suggestion
        return builder.WithEnvironment(context =>
        {
            var connectionStringValue = connectionString.Value ??
                builder.ApplicationBuilder.Configuration.GetConnectionString(connectionString.Name);

            if (string.IsNullOrEmpty(connectionStringValue))
            {
                throw new DistributedApplicationException($"A connection string for '{connectionString.Name}' could not be retrieved.");
            }

            if (builder.Resource is ContainerResource)
            {
                connectionStringValue = HostNameResolver.ReplaceLocalhostWithContainerHost(connectionStringValue, builder.ApplicationBuilder.Configuration);
            }

            context.EnvironmentVariables[$"{ConnectionStringEnvironmentName}{connectionString.Name}"] = connectionStringValue;
```

You are already capturing `connectionString` into the lambda, so you might as well make this name in the lambda instead of making the capture object bigger.

Reply:
I prefer to have the exact same code in main and preview2.

Reply:
Like I said above "My 2 comments can be made in main, we don't have to block the backport PR for them.".